cpp/bench/sg/umap.cu

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuda_utils.h>
+#include <cuml/cuml.hpp>
+#include <cuml/manifold/umap.hpp>
+#include <utility>
+#include "benchmark.cuh"
+
+namespace ML {
+namespace Bench {
+namespace umap {
+
+struct Params {
+  DatasetParams data;
+  BlobsParams blobs;
+  UMAPParams umap;
+};
+
+template <typename OutT, typename InT, typename IdxT>
+__global__ void castKernel(OutT* out, const InT* in, IdxT len) {
+  auto tid = IdxT(blockIdx.x) * blockDim.x + IdxT(threadIdx.x);
+  if (tid < len) {
+    out[tid] = OutT(in[tid]);
+  }
+}
+template <typename OutT, typename InT, typename IdxT = int>
+void cast(OutT* out, const InT* in, IdxT len, cudaStream_t stream) {
+  static const int TPB = 256;
+  auto nblks = MLCommon::ceildiv<IdxT>(len, TPB);
+  castKernel<OutT, InT, IdxT><<<nblks, TPB, 0, stream>>>(out, in, len);
+  CUDA_CHECK(cudaGetLastError());
+}
+
+class Umap : public BlobsFixture<float, int> {
+ public:
+  Umap(const std::string& name, const Params& p)
+    : BlobsFixture<float, int>(p.data, p.blobs), uParams(p.umap) {
+    this->SetName(name.c_str());
+  }
+
+ protected:
+  void runBenchmark(::benchmark::State& state) override {
+    if (!this->params.rowMajor) {
+      state.SkipWithError("Umap only supports row-major inputs");
+    }
+    auto& handle = *this->handle;
+    auto stream = handle.getStream();
+    cast<float, int>(yFloat, this->data.y, this->params.nrows, stream);
+    CUDA_CHECK(cudaStreamSynchronize(stream));
+    for (auto _ : state) {
+      CudaEventTimer timer(handle, state, true, stream);
+      fit(handle, this->data.X, yFloat, this->params.nrows, this->params.ncols,
```

Comment:
It's going to be important that we also benchmark the unsupervised `fit()` (the training code path without labels) and the `transform()` functions since all of those require subtle variations with the same optimizer. This will be important because if we're going to be making changes to the optimizer to speed up the execution, we're going to also want to consider the effects those changes have on the other code paths.

Reply:
That's a valid point! I had missed the various use-cases in here. Thanks for pointing them out. I have added separate suites for each of supervised/unsupervised/transform use-cases. Can you check now?